server/devdocs/bin/generate-components-usage-counts

```diff
@@ -0,0 +1,174 @@
+ */
+
+var async = require( 'async' ),
+	fs = require( 'fs' ),
+	fspath = require( 'path' ),
+	root = fspath.dirname( fspath.join( __dirname, '..', '..' ) ),
+
+	// Copyright (c) 2014-present, Facebook, Inc. See CREDITS.md#L1383
+	replacePatterns = {
+		BLOCK_COMMENT_RE: /(?:\/\*(.|\n)*?\*\/)|(\/\/.+(\n|$))/g,
+		LINE_COMMENT_RE: /\/\/.+(\n|$)/g,
+		IMPORT_RE: /(\bimport\s+?(?:.+\s+?from\s+?)?)(['"])([^'"]+)(\2)/g,
+		EXPORT_RE: /(\bexport\s+(?:[^'"]+\s+from\s+)??)(['"])([^'"]+)(\2)/g,
+		REQUIRE_RE: /(\brequire\s*?\(\s*?)(['"])([^'"]+)(\2\s*?\))/g
+	};
+
+function main() {
+	// extract list of files to index and remove leading ./'s
+	var fileList = process.
+		argv.
+		splice( 2, process.argv.length ).
+		map( function ( fileWithPath ) {
+			return fileWithPath.replace( /^\.\//, '' );
+		} );
+
+	if ( fileList.length === 0 ) {
+		process.stderr.write( 'You must pass a list of files to process (try "make server/devdocs/usage-counts.js"' );
+		process.exit( 1 );
+	}
+
+	getModulesWithDependencies( root, fileList )
+		.then(  function ( modulesWithDependencies ) {
+			var stats = generateStats( modulesWithDependencies );
+			writeStats( stats, 'server/devdocs/usage-counts.js' );
+			process.exit( 0 );
+		} )
+		.catch( function ( error ) {
+			console.error( "An error occurred while processing the files: \n\t", error );
+			process.exit( 1 );
+		} );
+}
+
+/**
+ * Calculate the modules dependencies.
+ * Produces an object that has the following structure:
+ * {
+ *  'A': [dep1, dep2, dep3, ...],
+ *  'B': [dep1, dep2, dep3, ...]
+ * }
+ * @param {string} root BasePath to resolve the files' path
+ * @param {Array} fileList A list of file names
+ * @returns {Promise}
+ */
+
+function getModulesWithDependencies( root, fileList ) {
+	return new Promise( function ( resolve, reject ) {
+		var filesContent = {},
+		dependencies = {};
+
+		async.each(
+			fileList,
+			function ( fileWithPath, next ) {
+				fs.readFile(
+					fspath.join( root, fileWithPath ),
+					{ encoding: 'utf8' },
+					function ( err, data ) {
+						if ( err ) {
+							return next( err );
+						}
+						// get the dependencies
+						filesContent[ fileWithPath ] = data;
+						next( null );
+					}
+				);
+			},
+			function ( err ) {
+				if ( err ) {
+					return reject( err );
+				}
+				Object.keys( filesContent ).reduce(function ( target, module ) {
+					var deps = getDependencies( filesContent[ module ] );
+					if ( deps.length ) {
+						target[ module ] = deps;
+					}
+					return target;
+				} , dependencies );
+
+				return resolve( dependencies );
+			}
+		);
+	} );
+}
+
+/**
+ * Extract all required modules from a `code` string.
+ * Copyright (c) 2014-present, Facebook, Inc. See CREDITS.md#L1383
+ */
+function getDependencies( code ) {
+	var cache = Object.create(null);
+	var deps = [];
+
+	function addDependency( dep ) {
+		if ( !cache[ dep ] ) {
+			cache[ dep ] = true;
+			deps.push(dep);
+		}
+	}
+
+	code
+		.replace( replacePatterns.BLOCK_COMMENT_RE, '' )
```

Comment:
Yeah fb has decent test coverage for them. See [current test suite](https://github.com/facebook/node-haste/blob/master/src/lib/__tests__/extractRequires-test.js) and [old test suite](https://github.com/facebook/react-native/blob/3a714c3eb8994ec515ca87ff6a86f3836225338f/packager/react-packager/src/DependencyResolver/haste/__tests__/HasteDependencyResolver-test.js).

<blockquote>
and we don’t need to properly parse the code?
</blockquote>

Yes that's indeed the right way to do so, build an AST of each file and grab the dependencies.
I did that in the first revision but it is crazy slow (3.x seconds). Static analysis is much faster, Facebook is using it to create a dependencies graph in [React Native Packager](https://github.com/facebook/react-native/tree/master/packager) (a tool similar to webpack but faster).